File: src/main/rdp/RdpLauncher.ts
import { app } from 'electron'
import { spawn, execFile } from 'child_process'
import { writeFileSync } from 'fs'
import { join } from 'path'
import type { Host } from '../../shared/types'
import type { VaultStore } from '../vault/VaultStore'

/**
 * Harici RDP istemcisine delege (1Remote "runner" deseni, rapor §8/1):
 * in-app WASM istemcisi NLA/çoklu ekran/HiDPI için yetersizse platformun
 * yerel RDP istemcisini başlatır.
 *  - Windows: cmdkey ile kimlik + mstsc
 *  - macOS: geçici .rdp dosyası + Microsoft Remote Desktop (parola kullanıcıdan)
 *  - Linux: xfreerdp / freerdp
 */
export class RdpLauncher {
  constructor(private vault: VaultStore) {}

  launch(host: Host): { ok: boolean; error?: string } {
    const identityId = this.vault.resolveIdentityId(host)
    const secrets = identityId ? this.vault.resolveSecrets(identityId) : null
    const username = secrets?.identity.username
    const password = secrets?.password
    const port = host.port || 3389

    try {
      switch (process.platform) {
        case 'win32':
          this.launchWindows(host.hostname, port, username, password)
          break
        case 'darwin':
          this.launchMac(host, port, username)
          break
        default:
          this.launchLinux(host.hostname, port, username, password)
      }
      return { ok: true }
    } catch (err) {
      return { ok: false, error: err instanceof Error ? err.message : String(err) }
    }
  }

  private launchWindows(
    hostname: string,
    port: number,
    username?: string,
    password?: string
  ): void {
    const target = `${hostname}:${port}`
    // Kimlik bilgisini geçici olarak Windows credential store'a yaz
    if (username && password) {
      execFile('cmdkey', [`/generic:TERMSRV/${hostname}`, `/user:${username}`, `/pass:${password}`])
    }
    const child = spawn('mstsc.exe', [`/v:${target}`], { detached: true, stdio: 'ignore' })
    child.unref()
    // Bağlantı kurulduktan sonra kimliği temizle (kısa gecikme)
    if (username) {
      setTimeout(() => execFile('cmdkey', [`/delete:TERMSRV/${hostname}`]), 20000)
    }
  }

  private launchMac(host: Host, port: number, username?: string): void {
    // macOS RDP istemcileri CLI ile parola almaz; .rdp dosyası üretip açıyoruz
    const lines = [
      `full address:s:${host.hostname}:${port}`,
      username ? `username:s:${username}` : '',
      'screen mode id:i:2',
      'authentication level:i:2'
    ].filter(Boolean)
    const filePath = join(app.getPath('temp'), `connexa-${host.id || 'rdp'}.rdp`)
    writeFileSync(filePath, lines.join('\n'), 'utf-8')
    const child = spawn('open', [filePath], { detached: true, stdio: 'ignore' })
    child.unref()
  }

  private launchLinux(
    hostname: string,
    port: number,
    username?: string,
    password?: string
  ): void {
    const args = [`/v:${hostname}`, `/port:${port}`]
    if (username) args.push(`/u:${username}`)
    if (password) args.push(`/p:${password}`)
    args.push('/cert:ignore', '+clipboard', '/dynamic-resolution')

    // xfreerdp yoksa freerdp'ye düş; ikisi de yoksa sessizce log'la
    // (spawn hatası asenkron geldiğinden launch() dönüşüne yansımaz)
    const tryClient = (client: string, onFail: () => void): void => {
      const child = spawn(client, args, { detached: true, stdio: 'ignore' })
      child.on('error', onFail)
      child.unref()
    }
    tryClient('xfreerdp', () => {
      tryClient('freerdp', () => {
        console.error('No FreeRDP client found (install xfreerdp/freerdp)')
      })
    })
  }
}

File: src-tauri/src/rdp_launcher.rs
//! Harici RDP istemcisi başlatıcı.
//! Platforma göre mstsc / open .rdp / xfreerdp.

use crate::types::Host;

pub fn launch(host: &Host, username: &str, password: &str) -> Result<(), String> {
    // addr yalnız windows/macos kollarında kullanılır; linux xfreerdp'de host+port
    // ayrı geçer. Platforma göre kullanılmadığında uyarı vermesin.
    #[allow(unused_variables)]
    let addr = format!(
        "{}:{}",
        host.hostname,
        if host.port == 0 { 3389 } else { host.port }
    );

    #[cfg(target_os = "windows")]
    {
        use std::process::Command;
        let target = format!("TERMSRV/{}", host.hostname);
        let _ = Command::new("cmdkey")
            .arg(format!("/generic:{target}"))
            .arg(format!("/user:{username}"))
            .arg(format!("/pass:{password}"))
            .output();
        Command::new("mstsc.exe")
            .arg(format!("/v:{addr}"))
            .spawn()
            .map_err(|e| format!("mstsc başlatılamadı: {e}"))?;
        // 20 sn sonra kimlik bilgisini sil
        let target2 = target.clone();
        std::thread::spawn(move || {
            std::thread::sleep(std::time::Duration::from_secs(20));
            let _ = Command::new("cmdkey")
                .arg(format!("/delete:{target2}"))
                .output();
        });
        return Ok(());
    }

    #[cfg(target_os = "macos")]
    {
        use std::process::Command;
        let _ = username;
        let _ = password;
        let file = std::env::temp_dir().join(format!("connexa-{}.rdp", host.id));
        let content = format!(
            "full address:s:{addr}\nusername:s:{username}\nscreen mode id:i:2\nauthentication level:i:2\n"
        );
        std::fs::write(&file, content).map_err(|e| format!("rdp dosyası yazılamadı: {e}"))?;
        Command::new("open")
            .arg(&file)
            .spawn()
            .map_err(|e| format!("open başarısız: {e}"))?;
        return Ok(());
    }

    #[cfg(all(unix, not(target_os = "macos")))]
    {
        use std::process::Command;
        let args: Vec<String> = vec![
            format!("/v:{}", host.hostname),
            format!("/port:{}", if host.port == 0 { 3389 } else { host.port }),
            format!("/u:{username}"),
            format!("/p:{password}"),
            "/cert:ignore".into(),
            "+clipboard".into(),
            "/dynamic-resolution".into(),
        ];
        for bin in ["xfreerdp", "freerdp"] {
            if Command::new(bin).args(&args).spawn().is_ok() {
                return Ok(());
            }
        }
        return Err("xfreerdp/freerdp bulunamadı".into());
    }

    #[allow(unreachable_code)]
    Err("bu platformda desteklenmiyor".into())
}
